### PR TITLE
Opt-in parallel processing of all messages in SQS response.

### PR DESCRIPTION
**Pull request by jacobengel**
- Added config setting ParallelHandling, defaulted to false
- Switch parallel looping through SQS ``response.Messages`` on ParallelHandling setting

### CODE REVIEW

**Comment by bfriesen** [comment]
Looks good. 👍

File: Rock.Messaging.SQS/Configuration/SQSConfiguration.cs [hunk 1]
```diff
@@ -8,13 +8,15 @@ public SQSConfiguration()
         {
             MaxMessages = 3;
             AutoAcknowledge = true;
+            ParallelHandling = false;
```
Comment (derikwhittaker) [inline Rock.Messaging.SQS/Configuration/SQSConfiguration.cs]: Defaults to false, why set explicitly?

File: Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs [hunk 2]
```diff
@@ -103,71 +106,83 @@ private void DoStuff()
...
+                Action acknowledge =
+                    () =>
                     {
-                        var receiptHandle = message.ReceiptHandle;
+                        Exception deleteException = null;
+                        DeleteMessageResponse deleteResponse = null;
 
-                        Action acknowledge =
-                            () =>
+                        for (int i = 0; i < 3; i++)
```
Comment (derikwhittaker) [inline Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs]: New to this code, but what is '3'?  Seems like a magic number.  Could this be better described as a Const?

File: Rock.Messaging.SQS/Configuration/SQSConfiguration.cs (hunk 1 above)
Comment (jacobengel) [reply Rock.Messaging.SQS/Configuration/SQSConfiguration.cs]: My motivation around adding it explicitly was readability, but that's fair - I'll remove this.

File: Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs [hunk 3]
```diff
@@ -103,71 +106,83 @@ private void DoStuff()
...
+                        {
+                            try
                             {
-                                Exception deleteException = null;
-                                DeleteMessageResponse deleteResponse = null;
+                                deleteException = null;
+                                deleteResponse = null;
```
Comment (derikwhittaker) [inline Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs]: Seems we do not need to set this to null given the next line will set the value when the .OverAsync returns.  Am I missing something?

File: Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs (hunk 2 above)
Comment (jacobengel) [reply Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs]: This was just moved around, but I can make it a const 😄

File: Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs (hunk 3 above)
Comment (jacobengel) [reply Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs]: Not sure, this was also just moved around.  @bfriesen thoughts?

File: Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs (hunk 2 above)
Comment (derikwhittaker) [reply Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs]: I looked around and was not able to understand what 3 meant.  So making it a const would help the next dev.

Comment (jacobengel) [reply Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs]: Not 100% sure I pegged the motivations right, as I didn't write this loop, but I took a crack at interpreting this as a const (as well as the other loop bound).

Comment (bfriesen) [reply Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs]: It's a simple retry loop. If the call to aws fails, try again two more times.

File: Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs (hunk 3 above)
Comment (bfriesen) [reply Rock.Messaging.SQS/MQ/SQSQueueReceiver.cs]: Took me a few minutes to remember, but this is to support better error reporting. Here's the hypothetical scenario:

We're in the retry loop, trying up to three times. The first time around, it isn't successful (doesn't matter why). The second time around, calling `_sqs.DeleteMessageAsync` does not throw an exception, but the response does not have a status code of `OK`. Note that `deleteResponse` *has* a value at this point before we loop again. On the third time around, `_sqs.DeleteMessageAsync` throws an exception, and it never assigns to `deleteResponse`. If assigning to null doesn't happen immediately before, then the error we report has mismatched error information and would be confusing - the exception would describe what went wrong when making the call to SQS, while the additional information from `deleteResponse` would describe a successful call to SQS with an unsuccessful result.

**Comment by bfriesen** [comment]
👍

**Pull request merged**